packages/ens-utils/src/ethregistrar.ts

```diff
@@ -3,323 +3,351 @@ import {
-  constructor(public contract: ContractRef) {
-    this.contract = contract;
+    return true;
   }
 
-  getTokenId(name: ENSName, isWrapped: boolean): TokenId {
-    if (!this.isClaimable(name, isWrapped)) {
+  public getRegistrationPriceQuote(
+    name: ENSName,
+    atTimestamp: Timestamp,
+    duration: Duration,
+    existingRegistration?: Registration,
+  ): RegistrationPriceQuote {
+    if (!this.canRegister(name, atTimestamp, duration, existingRegistration)) {
+      // TODO: refine the way we handle these exception cases.
       throw new Error(
-        `Unwrapped tokenId for name: "${name.name}" is not claimable by registrar: ${this.contract.address} on chainId: ${this.contract.chain.chainId}`,
+        `Cannot build registration price quote for name: "${name.name}".`,
+      );
+    }
+
+    const rentalPeriod = buildTimePeriod(
+      atTimestamp,
+      addSeconds(atTimestamp, duration),
+    );
+
+    let charges = this.getDurationCharges(name, duration);
+
+    let temporaryPremium: RegistrarTemporaryFee | null = null;
+
+    if (existingRegistration) {
+      temporaryPremium = this.getTemporaryPremiumCharge(
+        name,
+        existingRegistration,
+        atTimestamp,
       );
+      if (temporaryPremium) {
+        charges = [...charges, temporaryPremium];
+      }
     }
-    return buildTokenId(BigInt(labelhash(name.labels[0])));
+
+    const totalPrice = addPrices(charges.map((charge) => charge.price));
+
+    return {
+      action: RegistrarAction.Register,
+      rentalPeriod,
+      charges,
+      totalPrice,
+    };
   }
 
-  isClaimable(name: ENSName, isWrapped: boolean): boolean {
-    // name must be unwrapped
-    if (isWrapped) return false;
+  public canRenew(
+    name: ENSName,
+    atTimestamp: Timestamp,
+    duration: Duration,
+    existingRegistration: Registration,
+  ): boolean {
+    if (!this.getManagedSubname(name)) {
+      // name is not directly managed by this registrar
+      return false;
+    }
 
-    // must have exactly 2 labels to be a direct subname of ".eth"
-    if (name.labels.length !== 2) return false;
+    if (existingRegistration) {
+      if (
+        existingRegistration.registrationTimestamp &&
+        atTimestamp.time < existingRegistration.registrationTimestamp.time
+      ) {
+        // if the renewal is requested before the registration, we can't renew it
+        // TODO: check for possible off-by-1 errors in the logic above
+        return false;
+      }
 
-    // last label must be "eth"
-    if (name.labels[1] !== ETH_TLD) return false;
+      const releaseTimestamp = getDomainReleaseTimestamp(existingRegistration);
 
-    // first label must be of sufficient length
-    return charCount(name.labels[0]) >= MIN_ETH_REGISTRABLE_LABEL_LENGTH;
+      if (releaseTimestamp && releaseTimestamp.time > atTimestamp.time) {
+        // if the name is released, we can't renew it anymore
+        // TODO: check for possible off-by-1 errors in the logic above
```

Comment:
@lightwalker-eth can you extend on off-by-1 spec?

Reply:
We need to see how ENS smart contracts actually work. Ex: Actual expiration time? 1 second earlier? 1 second later?

Reply:
<img width="909" alt="Captura de Tela 2024-08-06 às 20 16 40" src="https://github.com/user-attachments/assets/b6fc2e19-8563-4e89-8e0c-fa7d6aed1ba5">

This function was extracted from BaseRegistrar and reveals that above logic could look different: added a commit suggestion above!